engines/scumm/dimuse_v2/dimuse_v2_files.cpp

```diff
@@ -0,0 +1,280 @@
+			return resultingOffset;
+		} 
+		debug(5, "DiMUSEFilesHandler::seek(): can't find sound %d (%s); did you forget to open it?", soundId, fileName);
+	} else {
+		debug(5, "DiMUSEFilesHandler::seek(): soundId is 0 or out of range");
+	}
+
+	return 0;
+}
+
+int DiMUSEFilesHandler::read(int soundId, uint8 *buf, int size, int bufId) {
+	// This function and files_seek() are used for sounds for which a stream is needed
+	// (speech and music), therefore they will always refer to sounds in a bundle file
+	// TODO: Does this work with speech?
+	if (soundId != 0 && soundId < 0xFFFFFFF0) {
+		char fileName[23] = "";
+		getFilenameFromSoundId(soundId, fileName);
+
+		DiMUSESndMgr::SoundDesc *s = _sound->getSounds();
+		DiMUSESndMgr::SoundDesc *curSnd = NULL;
+		for (int i = 0; i < MAX_IMUSE_SOUNDS; i++) {
+			curSnd = &s[i];
+			if (curSnd->inUse) {
+				if (curSnd->soundId == soundId) {
+					uint8 *tmpBuf;
+					//debug(5, "DiMUSE_v2::files_read(): trying to read (%d) bytes of data from file %s", size, fileName);
+					int resultingSize = curSnd->bundle->readFile(fileName, size, &tmpBuf, ((_vm->_game.id == GID_CMI) && !(_vm->_game.features & GF_DEMO)));
+
+					if (resultingSize != size)
+						debug(5, "DiMUSEFilesHandler::read(): WARNING: tried to read %d bytes, got %d instead (soundId %d (%s))", size, resultingSize, soundId, fileName);
+					memcpy(buf, tmpBuf, size);
+					free(tmpBuf);
+					return resultingSize;
+				}
+			}
+		}
+
+		debug(5, "DiMUSEFilesHandler::read(): can't find sound %d (%s); did you forget to open it?", soundId, fileName);
+
+	} else {
+		debug(5, "DiMUSEFilesHandler::read(): soundId is 0 or out of range");
+	}
+	
+	return 0;
+}
+
+	
+DiMUSESoundBuffer *DiMUSEFilesHandler::getBufInfo(int bufId) {
+	if (bufId > 0 && bufId <= 4) {
+		return &_soundBuffers[bufId];
+	}
+
+	debug(5, "DiMUSEFilesHandler::getBufInfo(): ERROR: invalid buffer id");
+	return NULL;
+}
+
+void DiMUSEFilesHandler::openSound(int soundId) {
+	char fileName[23] = "";
+	getFilenameFromSoundId(soundId, fileName);
+	DiMUSESndMgr::SoundDesc *s = NULL;
+	int groupId = soundId == kTalkSoundID ? IMUSE_VOLGRP_VOICE : IMUSE_VOLGRP_MUSIC;
+	s = _sound->openSound(soundId, fileName, IMUSE_BUNDLE, groupId, -1);
+	if (!s)
+		s = _sound->openSound(soundId, fileName, IMUSE_BUNDLE, groupId, 1);
+	if (!s)
+		s = _sound->openSound(soundId, fileName, IMUSE_BUNDLE, groupId, 2);
+	if (!s)
+		debug(5, "DiMUSEFilesHandler::openSound(): can't open sound %d (%s)", soundId, fileName);
+}
+
+void DiMUSEFilesHandler::closeSound(int soundId) {
+	_sound->scheduleSoundForDeallocation(soundId);
+}
+
+void DiMUSEFilesHandler::closeAllSounds() {
+	DiMUSESndMgr::SoundDesc *s = _sound->getSounds();
+	for (int i = 0; i < MAX_IMUSE_SOUNDS; i++) {
+		if (s[i].inUse)
+			closeSound(s->soundId);
+	}
+
+	_engine->flushTracks();
+}
+
+void DiMUSEFilesHandler::getFilenameFromSoundId(int soundId, char *fileName) {
+	int i = 0;
+
+	if (soundId == kTalkSoundID) {
+		strcpy(fileName, _currentSpeechFile);
```

Comment:
buffer size checks here and the at the following locations. This should most likely get a `filenameSize` parameter or should be a `Common::String` if possible

Reply:
What if I handled it like this?
```c++
void DiMUSEFilesHandler::getFilenameFromSoundId(int soundId, char *fileName) {
	int i = 0;

	if (soundId == kTalkSoundID) {
		strncpy(fileName, _currentSpeechFile, 60);
	}
...
}
```

This function is only ever called with `fileName` being a `char[60]` (originally 12, I just changed it to 60, to keep only one arbitrary size).

Reply:
i would still vote for doing

```
void DiMUSEFilesHandler::getFilenameFromSoundId(int soundId, char *fileName, size_t size) {
  [...]
}
```

and call with

```
getFilenameFromSoundId(soundId, filename, sizeof(filename));
```

the other approach is quite error prone if something is changing in the future.

And please also see the needed null termination from above:

```
strncpy(filename, name, size - 1);
filename[size - 1] = '\0';
```

Reply:
I'm sure scummvm has a proper string copy method that ensures the null termination and takes the buffer size into account.

Reply:
Common::strlcpy

Reply:
Thanks! :)